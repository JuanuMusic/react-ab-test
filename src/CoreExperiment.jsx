import React, { useState, useMemo } from "react";
import PropTypes from "prop-types";
import useExperiment from "./hooks";
import emitter from "./emitter";

const filterVariants = (name, children) => {
  let variants = {};
  React.Children.forEach(children, (element) => {
    if (
      !React.isValidElement(element) ||
      element.type.displayName !== "Pushtell.Variant"
    ) {
      let error = new Error(
        "Pushtell Experiment children must be Pushtell Variant components."
      );
      error.type = "PUSHTELL_INVALID_CHILD";
      throw error;
    }
    variants[element.props.name] = element;
    emitter.addExperimentVariant(name, element.props.name);
  });
  emitter.emit("variants-loaded", name);
  return variants;
};

const CoreExperiment = (props) => {
  const variants = useMemo(() => {
    return filterVariants(props.name, props.children);
  }, [props.name, props.children]);

  const { selectVariant } = useExperiment(
    props.name,
    props.userIdentifier,
    props.defaultVariantName
  );

  return selectVariant(variants, []);
};

export default CoreExperiment;
