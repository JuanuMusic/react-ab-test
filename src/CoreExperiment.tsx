import React, { useMemo } from 'react';
import useExperiment from './hook';
import emitter from './emitter';
import { ExperimentProps, VariantMap } from './types';

interface ErrorWithType extends Error {
  type?: string;
}

const filterVariants = (
  name: string,
  children: React.ReactNode
): VariantMap => {
  const variants: VariantMap = {};

  React.Children.forEach(children, (element) => {
    if (
      !React.isValidElement(element) ||
      !element.type ||
      typeof element.type === 'string' ||
      (element.type as React.ComponentType).displayName !== 'Pushtell.Variant'
    ) {
      const error: ErrorWithType = new Error(
        'Pushtell Experiment children must be Pushtell Variant components.'
      );
      error.type = 'PUSHTELL_INVALID_CHILD';
      throw error;
    }
    variants[element.props.name] = element;
    emitter.addExperimentVariant(name, element.props.name);
  });

  emitter.emit('variants-loaded', name);
  return variants;
};

const CoreExperiment = (props: ExperimentProps): React.ReactElement => {
  const variants = useMemo(() => {
    return filterVariants(props.name, props.children);
  }, [props.name, props.children]);

  const { selectVariant } = useExperiment(
    props.name,
    props.userIdentifier,
    props.defaultVariantName
  );

  return selectVariant(
    variants,
    React.createElement(React.Fragment)
  ) as React.ReactElement;
};

export default CoreExperiment;
