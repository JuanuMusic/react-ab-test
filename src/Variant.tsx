import React from 'react';
import { VariantProps } from './types';

export function Variant(props: VariantProps) {
  if (React.isValidElement(props.children)) {
    return props.children;
  } else {
    return <span>{props.children}</span>;
  }
}

Variant.displayName = 'Pushtell.Variant';

export default Variant;
