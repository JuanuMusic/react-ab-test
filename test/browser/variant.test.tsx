import React from 'react';
import { renderWithWrapper } from '../test-utils';

import CoreExperiment from '../../src/CoreExperiment';
import Variant from '../../src/Variant';

describe('Variant', () => {
  it('should render text nodes', () => {
    const variantTextA = 'variantTextA';
    const variantTextB = 'variantTextB';

    renderWithWrapper(
      <CoreExperiment name="text-nodes" defaultVariantName="A">
        <Variant name="A">{variantTextA}</Variant>
        <Variant name="B">{variantTextB}</Variant>
      </CoreExperiment>
    );

    // Since CoreExperiment only renders the active variant (A),
    // we should find its content in the document
    const content = document.body.textContent;
    expect(content).toContain(variantTextA);
    expect(content).not.toContain(variantTextB);
  });

  it('should render components', () => {
    const wrapper = renderWithWrapper(
      <CoreExperiment name="components" defaultVariantName="A">
        <Variant name="A">
          <div id="variant-a" />
        </Variant>
        <Variant name="B">
          <div id="variant-b" />
        </Variant>
      </CoreExperiment>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render arrays of components', () => {
    const wrapper = renderWithWrapper(
      <CoreExperiment name="array-of-elements" defaultVariantName="A">
        <Variant name="A">
          <div id="variant-a" />
          <div />
        </Variant>
        <Variant name="B">
          <div id="variant-b" />
          <div />
        </Variant>
      </CoreExperiment>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
