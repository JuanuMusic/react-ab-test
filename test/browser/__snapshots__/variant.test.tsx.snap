// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variant should render arrays of components 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span>
        <div
          id="variant-a"
        />
        <div />
      </span>
    </div>
  </body>,
  "container": <div>
    <span>
      <div
        id="variant-a"
      />
      <div />
    </span>
  </div>,
  "debug": [Function],
  "find": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "setProps": [Function],
  "unmount": [Function],
}
`;

exports[`Variant should render components 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="variant-a"
      />
    </div>
  </body>,
  "container": <div>
    <div
      id="variant-a"
    />
  </div>,
  "debug": [Function],
  "find": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "setProps": [Function],
  "unmount": [Function],
}
`;
